File: apps/mqtt-worker/src/parser.ts
import { z } from 'zod';
import { Parameter, Protocol, ReadingQuality } from '@moondesk/domain';

/**
 * Expected MQTT message format from edge devices
 * Topic format: moondesk/{organizationId}/sensors/{sensorId}/readings
 */
export const mqttReadingSchema = z.object({
    // Sensor identification
    sensorId: z.number(),

    // Reading data
    value: z.number(),
    timestamp: z.string().datetime().optional(),

    // Optional metadata
    parameter: z.nativeEnum(Parameter).optional(),
    quality: z.nativeEnum(ReadingQuality).optional(),
    metadata: z.record(z.string()).optional(),
});

export type MqttReading = z.infer<typeof mqttReadingSchema>;

/**
 * Batch message for multiple readings
 */
export const mqttBatchSchema = z.object({
    readings: z.array(mqttReadingSchema),
});

export type MqttBatch = z.infer<typeof mqttBatchSchema>;

/**
 * Alert threshold configuration message
 */
export const mqttAlertConfigSchema = z.object({
    sensorId: z.number(),
    minValue: z.number().optional(),
    maxValue: z.number().optional(),
    enabled: z.boolean().default(true),
});

export type MqttAlertConfig = z.infer<typeof mqttAlertConfigSchema>;

/**
 * Command response from edge device
 */
export const mqttCommandResponseSchema = z.object({
    commandId: z.number(),
    status: z.enum(['completed', 'failed', 'timeout']),
    result: z.string().optional(),
    error: z.string().optional(),
});

export type MqttCommandResponse = z.infer<typeof mqttCommandResponseSchema>;

/**
 * Parse topic to extract organization and sensor IDs
 * Format: moondesk/{organizationId}/sensors/{sensorId}/{action}
 */
export interface ParsedTopic {
    organizationId: string;
    sensorId: number;
    action: 'readings' | 'batch' | 'status' | 'command-response';
}

export function parseTopic(topic: string): ParsedTopic | null {
    const parts = topic.split('/');

    // Expected: moondesk/{orgId}/sensors/{sensorId}/{action}
    if (parts.length < 5 || parts[0] !== 'moondesk' || parts[2] !== 'sensors') {
        return null;
    }

    const organizationId = parts[1];
    const sensorId = parseInt(parts[3] ?? '', 10);
    const action = parts[4] as ParsedTopic['action'];

    if (!organizationId || isNaN(sensorId)) {
        return null;
    }

    const validActions = ['readings', 'batch', 'status', 'command-response'];
    if (!validActions.includes(action)) {
        return null;
    }

    return { organizationId, sensorId, action };
}

/**
 * Parse and validate MQTT message payload
 */
export function parseMessage(
    payload: Buffer,
    action: ParsedTopic['action']
): MqttReading | MqttBatch | MqttCommandResponse | null {
    try {
        const data = JSON.parse(payload.toString());

        switch (action) {
            case 'readings':
                return mqttReadingSchema.parse(data);
            case 'batch':
                return mqttBatchSchema.parse(data);
            case 'command-response':
                return mqttCommandResponseSchema.parse(data);
            default:
                return null;
        }
    } catch (error) {
        return null;
    }
}

File: internal/worker/parser.go
package worker

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"

	"github.com/abdssamie/moondesk-go/internal/domain"
)

type MqttReading struct {
	SensorID  int64             `json:"sensorId"`
	Value     float64           `json:"value"`
	Timestamp string            `json:"timestamp,omitempty"`
	Parameter domain.Parameter  `json:"parameter,omitempty"`
	Quality   domain.ReadingQuality `json:"quality,omitempty"`
	Metadata  map[string]string `json:"metadata,omitempty"`
}

type MqttBatch struct {
	Readings []MqttReading `json:"readings"`
}

type MqttCommandResponse struct {
	CommandID int64  `json:"commandId"`
	Status    string `json:"status"`
	Result    string `json:"result,omitempty"`
	Error     string `json:"error,omitempty"`
}

type ParsedTopic struct {
	OrganizationID string
	SensorID       int64
	Action         string // readings, batch, status, command-response
}

func ParseTopic(topic string) (*ParsedTopic, error) {
	parts := strings.Split(topic, "/")

	// Expected: moondesk/{orgId}/sensors/{sensorId}/{action}
	if len(parts) < 5 || parts[0] != "moondesk" || parts[2] != "sensors" {
		return nil, fmt.Errorf("invalid topic format")
	}

	orgID := parts[1]
	sensorID, err := strconv.ParseInt(parts[3], 10, 64)
	if err != nil {
		return nil, fmt.Errorf("invalid sensor id in topic")
	}

	action := parts[4]

	return &ParsedTopic{
		OrganizationID: orgID,
		SensorID:       sensorID,
		Action:         action,
	}, nil
}

func ParseMessage(payload []byte, action string) (interface{}, error) {
	switch action {
	case "readings":
		var msg MqttReading
		if err := json.Unmarshal(payload, &msg); err != nil {
			return nil, err
		}
		return msg, nil
	case "batch":
		var msg MqttBatch
		if err := json.Unmarshal(payload, &msg); err != nil {
			return nil, err
		}
		return msg, nil
	case "command-response":
		var msg MqttCommandResponse
		if err := json.Unmarshal(payload, &msg); err != nil {
			return nil, err
		}
		return msg, nil
	default:
		return nil, fmt.Errorf("unhandled action: %s", action)
	}
}
